File: algorithm/tree/Leetcode-1038-BinarySearchTreetoGreaterSumTree/Leetcode-1038-BinarySearchTreetoGreaterSumTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    sum_of_node = 0
    def bstToGst(self, root: TreeNode) -> TreeNode:
        discovered = set()
        def dfs(node) :
            if not node : return 

            dfs(node.right)
            if node.val not in discovered :
                discovered.add(node.val)
                self.sum_of_node += node.val
                node.val = self.sum_of_node
                discovered.add(node.val)
            else : pass
            dfs(node.left)

            return
        dfs(root)
        return root

File: 2024상반기/5월 1주차/leetcode-1038-Binary Search Tree to Greater Sum Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rounding(TreeNode* node, int n) {
        if (node == nullptr) return n;
        int rsum = rounding(node->right, n);
        node->val += rsum;
        int lsum = rounding(node->left, node->val);
        return lsum;
    }
    TreeNode* bstToGst(TreeNode* root) {
        rounding(root, 0);
        return root;
    }
};